Make Tapir more generic and configurable for Non-SuperCoop-Berlin-Users

fabiangebert opened the pull request:
Hey Tapir Team! I've created some patches to make Tapir more generic and configurable for Non-SuperCoop-Berlin-Users. Also I've fixed some minor things.
I have hopefully split the changes in comprehensible commits.
Looking forward to your feedback.
Thanks for all your work, it's great to have Tapir!

fabiangebert commented:
> Hi! Thanks a lot for contributing. I'm not used to reviewing PRs, please bear with me :D Overall I think the changes are good, I have a few points that we could discuss.

Sure thing! I think you can create a review here on GitHub to show me your points or add a comment here - whatever works better for you.

Theophile-Madet commented:
Hey, yeah sorry I was still writing my comment and published it too early, it's updated now.

fabiangebert commented:
> Hi! Thanks a lot for contributing. I'm not used to reviewing PRs, please bear with me :D Overall I agree with the changes, I have a few points that we could discuss. -My first thought on configuration through environment variables was that it leaves stuff undocumented, but I guess they belong in the docker-compose.yml that is used for deployment? -Translation typo fixes are great. You also slightly rephrased a few things, have you had that checked by a native speaker or someone very confident in their english?

I've added the respective environment variables to our deployment docker-compose.yml:
```
      SITE_URL: "https://....wirmarkt.de"
      SECRET_KEY: ${TAPIR_SECRET_KEY}
      EMAIL_ENV: "prod"
      EMAIL_HOST: ...
      EMAIL_HOST_USER: ...
      EMAIL_HOST_PASSWORD: ${TAPIR_SMTP_PASSWORD}
      DJANGO_ADMINS: "WirMarkt Office <...>"
      DATABASE_URL: "postgresql://tapir:${TAPIR_POSTGRES_PASSWORD}@db/tapir"
      LDAP_URL: "ldap://cn=admin,dc=wirmarkt,dc=de:${LDAP_ADMIN_PASSWORD}@ldap"
      EMAIL_ADDRESS_MEMBER_OFFICE: ...
      SERVER_EMAIL: ...
      REG_PERSON_BASE_DN: "ou=people,dc=wirmarkt,dc=de"
      REG_GROUP_BASE_DN: "ou=groups,dc=wirmarkt,dc=de"
      LDAP_WELCOME_DESK_ID: "CN=welcome-desk.members.wirmarkt.de,O=WirMarkt,C=DE"
```
We could simply add some comments to settings.py to explain the variables?

As for the translations: I have changed the translations that I am sure are wrong or at least very strange. I can ask a friend who is English for advice on specific translations if you like. Feel free to highlight in the commit translations that you'd like me to double check.

leonhandreke commented:
Thanks a lot! I cherry-picked the low-hanging fruit to master using the old way (tm) to already make the patchset a bit smaller and easier to review here. Feel free to rebase (`git rebase master && git push --force` should be enough).

fabiangebert commented:
@leonhandreke done that, should be correct now.

leonhandreke commented on tapir/settings.py:
```diff
@@ -198,9 +201,9 @@
 WEASYPRINT_BASEURL = "/"
 
-REG_PERSON_BASE_DN = "ou=people,dc=supercoop,dc=de"
+REG_PERSON_BASE_DN = env("REG_PERSON_BASE_DN", default="ou=people,dc=supercoop,dc=de")
```
Same here, if this does not differ between envs, I feel it's probably better to keep it in settings.py for now.

fabiangebert replied:
OK, if we go for a everyone-should-fork-strategy, it makes perfect sense without the env-var.

leonhandreke commented on tapir/coop/templates/coop/membership_agreement_pdf.default.html:
```diff
@@ -0,0 +1,259 @@
+<span style="border-top: 1px solid grey; width: 5mm; position: fixed; left: -15mm; top: 60mm;"></span>
+<span style="border-top: 1px solid grey; width: 5mm; position: fixed; left: -15mm; top: 165mm;"></span>
+
+<!-- A DIN 5008 Letter has 2,7cm of top margin -->
```
If we're having multiple of these now, maybe it makes sense to have a DIN 5008 letter in the core and have `{% block %}` placeholders to fill in the content?

No blocker, feel free to merge like this, we'll clean it up later.

fabiangebert replied:
Yes, absolutely!

leonhandreke commented on tapir/coop/templates/coop/membership_agreement_pdf.default.html:
```diff
@@ -0,0 +1,259 @@
+</div>
+
+<div style="text-align: center; font-size: 14pt; font-weight: bold;">
+    Beteiligungserklärung <em>Organisation</em>
```
Should we put COOP_NAME here? Or do we not expect anybody to use this anyway and instead keep it as a template?

If we expect the default template to not be used in any production setting anyway, let's put a comment in the rendering code and the template to indicate that this template is really only expected to be rendered in non-customized "fresh" versions of the code and should always be adapted per Coop.

fabiangebert replied:
Yes, I've but in placeholders to make people completely adapt the whole template.
I'll add the remarks!

leonhandreke reviewed (commented):
Thanks for submitting this! See a few inline comments.

fabiangebert commented on tapir/settings.py:
```diff
@@ -168,22 +167,25 @@
+EMAIL_ADDRESS_MEMBER_OFFICE = env(
+    "EMAIL_ADDRESS_MEMBER_OFFICE", default="mitglied@supercoop.de"
+)
+COOP_NAME = env("COOP_NAME", default="SuperCoop Berlin")
```
Interesting thoughts!
What makes COOP_NAME interesting is that you can merge changes (in particular i18n-related stuff in the .po files) a lot easier.
It could also make a docker-compose-only deployment possible, where you put your custom templates in volumes and define the coop name in env?
That might be too far fetched, though.

leonhandreke replied:
Yes, I totally agree we should have `COOP_NAME` and many other vars! My suggestion would be to keep them in `settings.py` though and not set them through the environment.

My feeling is that new deployments only through the env aren't realistic, as every coop will probably require some small code change somewhere. I'm not even sure it's desirable: Having a git history of template and policy changes and building them all into the image is quite desirable I think. Also makes rollbacks easier in case something breaks, since almost all config is baked into the Docker image.

fabiangebert commented on tapir/shifts/templates/shifts/shift_detail.html:
```diff
@@ -163,15 +163,17 @@ <h5 class="d-flex justify-content-between align-items-center">
                                                 -The shift is in the future<br/>
                                             {% endblocktranslate %}
-                                            <span {% if not slot.can_register %} data-bs-toggle="tooltip" data-html="true"
+                                        {% autoescape off %}
```
Otherwise, Bootstrap will render python escape chars and not the HTML content.

leonhandreke replied:
Ah, now I get it.

Seems like a garbage design decision by Bootstrap, turning off autoescaping is potentially a dangerous move in every framework... but obviously fine in this case since we're not dealing with user input :)

fabiangebert commented:
I think there are two major open points to be discussed (the other's I'll deal with in another commit to this PR)
* should every coop fork or would a docker-compose deployment from a common image with some variables and templates be desirable? (See answers to @leonhandreke's inline comments)
* what language-specific changes exactly would you like me to double-check @Theophile-Madet?

Theophile-Madet commented:
On the translations, you don't need to get anything checked further, I was just hoping that we finally got a trained person to check our wording :p

On the env-variables, I also find clearer if they can be in the same repo, but I'd try as much as possible to make the org-specific changes separate from the core stuff. Would it work to have settings.py import a settings_org.py at the end, where each org put their overrides?

leonhandreke commented on pyproject.toml:
```diff
@@ -6,34 +6,34 @@ authors = ["Leon Handreke <leonh@ndreke.de>"]
 
```
Totally agree with everything you said. Also in favor of keeping everything as up to date as possible.

I'd be interested what the command was to generate this change. I personally wouldn't know how to, but I'm not a poetry expert either. @fabiangebert :) ?

fabiangebert replied:
There's a package called `poetryup` that does that!

leonhandreke replied:
Thanks! Not that bothered about it being in this PR, would be better to keep it separate in the future, but let's try to reduce work here as much as possible :)

leonhandreke commented:
@Theophile-Madet regarding env vars: Don't have a strong opinion here, and definitely no problem to import an `org_settings.py`. However, keeping them in one file might also have advantages. If the config file line is changed in the common forkable core (i.e. because we changed the setting slightly), you are explicitly requested to merge. Having a separate file would not achieve this.

Debian does it a similar way, where if a package distributes a new default config file, you're always requested to merge. Not sure what the rationale here is, but maybe similar. If something in the config has changed drastically, you'll notice.

But as said, no very strong opinions here.

fabiangebert commented:
I hope to have addressed all outstanding issues - if not, let me know.
I've seen you've added SonarCloud and it requires a `SONAR_TOKEN` to be set. Is that sth. you would advise us to have in our fork as well or should I disable it?

Theophile-Madet commented:
We haven't finished the SonarCloud config yet, it's missing the test coverage reports. 
It's up to you if you want to have it. I think it's nice, it helps a bit with accessibility and it can spot some vulnerabilities. Also having being able to show that the test coverage is high could help motivate contributors. You can have a look here : https://sonarcloud.io/summary/overall?id=SuperCoopBerlin_tapir
